pkg/controller/service/servicecontroller.go

```diff
@@ -767,3 +741,83 @@ func (s *cachedService) nextRetryDelay() time.Duration {
 func (s *cachedService) resetRetryDelay() {
 	s.lastRetryDelay = time.Duration(0)
 }
+
+
+// syncService will sync the Service with the given key if it has had its expectations fulfilled,
+// meaning it did not expect to see any more of its pods created or deleted. This function is not meant to be
+// invoked concurrently with the same key.
+func (sc *ServiceController) syncService(key string) error {
+	startTime := time.Now()
+	var cachedService *cachedService
+	var retryDelay time.Duration
+	defer func() {
+		glog.V(4).Infof("Finished syncing service %q (%v)", key, time.Now().Sub(startTime))
+	}()
+	// obj holds the latest service info from apiserver
+	obj, exists, err := sc.serviceStore.Store.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		sc.queue.Add(key)
+		return err
+	}
+	if !exists {
+		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
+		glog.Infof("Service has been deleted %v", key)
+		err, retryDelay = sc.processServiceDeletion(key)
+	}
+	if exists {
+		service, ok := obj.(*api.Service)
+		if ok {
+			cachedService = sc.cache.getOrCreate(key)
+			err, retryDelay = sc.processServiceUpdate(cachedService, service, key)
```

Comment:
do we really need a cached service? we know what we want (service spec) and we know what exists (cloudprovider resources) can't we just sync them? I'm fine with a TODO, but the less local state this controller maintains, the more robust it's going to be on restart etc